### PR TITLE
fix: change permissions and remove files on rmtree

### PR DESCRIPTION
**Pull request by sriram-mv**
### fix: change permissions and remove files on rmtree

Why is this change necessary?

* sam init clones from a git repository called aws-sam-cli-app-templates, but when attempting to reclone, removal of the old directory becomes problematic due to the read-only nature of the `.git` files that get downloaded and results in a traceback.

How does it address the issue?

* If a given file is not able to be deleted due to permission issues, write permissions are added to the file in question and deleted.

What side effects does this change have?

* None, the previous side-effects of leaving behind .git objects will be addressed.

- [ ] Write Design Document ([Do I need to write a design document?](https://github.com/awslabs/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.rst#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samcli/lib/utils/osutils.py [hunk 1]
```diff
@@ -53,7 +50,11 @@ def mkdir_temp(mode=0o755, ignore_errors=False):
 
 
 def _rmtree_callback(function, path, excinfo):
```
Comment (c2tarun) [inline samcli/lib/utils/osutils.py]: I would recommend adding a docstring here, not as detailed as public methods, but a minor one explaining that this method will modify folder permission.

Comment (c2tarun) [reply samcli/lib/utils/osutils.py]: One more question, do you want to prefix this method name with `_` given that this method is imported in another module?

Comment (sriram-mv) [reply samcli/lib/utils/osutils.py]: Done! 👏

**Comment by rajendragosavi** [comment]
I see the issue still persists. I had to follow @sriram-mv  instructions manually. I am using sam 0.52.0 and I still see this issue.

**Comment by kannan-kiwitech** [comment]
I'm getting the same issue in recent version

SAM CLI, version 1.46.0 Windows 11

`2022-04-15 14:44:58,488 |
Cloning from https://github.com/aws/aws-sam-cli-app-templates (process may take a moment)
2022-04-15 14:45:02,933 | Removing old repo at C:\Users\xxx\AppData\Roaming\AWS SAM\aws-sam-cli-app-templates
2022-04-15 14:45:03,293 | Copying from C:\Users\xxx\AppData\Local\Temp\tmp_6asyom5\aws-sam-cli-app-templates to C:\Users\xxx\AppData\Roaming\AWS SAM\aws-sam-cli-app-templates
2022-04-15 14:45:07,097 | rmtree failed in <built-in function unlink> for C:\Users\xxx\AppData\Local\Temp\tmp_6asyom5\aws-sam-cli-app-templates\dotnetcore3.1\cookiecutter-aws-sam-quick-start-cloudwatch-events-dotnet\{{cookiecutter.project_name}}\test\{{cookiecutter.project_name}}.Tests\{{cookiecutter.project_name}}.Tests.csproj, details: (<class 'FileNotFoundError'>, FileNotFoundError(2, 'The system cannot find the path specified'), <traceback object at 0x000001B0E94FE380>)
2022-04-15 14:45:07,097 | rmtree failed in <built-in function rmdir> for C:\Users\xxx\AppData\Local\Temp\tmp_6asyom5\aws-sam-cli-app-templates\dotnetcore3.1\cookiecutter-aws-sam-quick-start-cloudwatch-events-dotnet\{{cookiecutter.project_name}}\test\{{cookiecutter.project_name}}.Tests, details: (<class 'OSError'>, OSError(41, 'The directory is not empty'), <traceback object at 0x000001B0E94FE380>)
2022-04-15 14:45:07,098 | rmtree failed in <built-in function rmdir> for C:\Users\xxx\AppData\Local\Temp\tmp_6asyom5\aws-sam-cli-app-templates\dotnetcore3.1\cookiecutter-aws-sam-quick-start-cloudwatch-events-dotnet\{{cookiecutter.project_name}}\test, details: (<class 'OSError'>, OSError(41, 'The directory is not empty'), <traceback object at 0x000001B0E94FE300>)
2022-04-15 14:45:07,098 | rmtree failed in <built-in function rmdir> for C:\Users\xxx\AppData\Local\Temp\tmp_6asyom5\aws-sam-cli-app-templates\dotnetcore3.1\cookiecutter-aws-sam-quick-start-cloudwatch-events-dotnet\{{cookiecutter.project_name}}, details: (<class 'OSError'>, OSError(41, 'The directory is not empty'), <traceback object at 0x000001B0E94FE2C0>)
2022-04-15 14:45:07,098 | rmtree failed in <built-in function rmdir> for C:\Users\xxx\AppData\Local\Temp\tmp_6asyom5\aws-sam-cli-app-templates\dotnetcore3.1\cookiecutter-aws-sam-quick-start-cloudwatch-events-dotnet, details: (<class 'OSError'>, OSError(41, 'The directory is not empty'), <traceback object at 0x000001B0E94FE280>)
2022-04-15 14:45:07,114 | rmtree failed in <built-in function rmdir> for C:\Users\xxx\AppData\Local\Temp\tmp_6asyom5\aws-sam-cli-app-templates\dotnetcore3.1, details: (<class 'OSError'>, OSError(41, 'The directory is not empty'), <traceback object at 0x000001B0E94EDF40>)
2022-04-15 14:45:07,374 | rmtree failed in <built-in function rmdir> for C:\Users\xxx\AppData\Local\Temp\tmp_6asyom5\aws-sam-cli-app-templates, details: (<class 'OSError'>, OSError(41, 'The directory is not empty'), <traceback object at 0x000001B0E94EDFC0>)
2022-04-15 14:45:07,374 | rmtree failed in <built-in function rmdir> for C:\Users\xxx\AppData\Local\Temp\tmp_6asyom5, details: (<class 'OSError'>, OSError(41, 'The directory is not empty'), <traceback object at 0x000001B0E94E96C0>)
2022-04-15 14:45:07,374 | Template is not provided in context, skip adding project type metric
2022-04-15 14:45:07,374 | Sending Telemetry: {'metrics': [{'commandRun': {'requestId': '5da8e467-86df-452a-97c9-a3e731b4bb19', 'installationId': '6faffe73-4122-48c5-99f4-d776ccc56cff', 'sessionId': 'c0934b02-bd88-459e-aae4-6332a4893a22', 'executionEnvironment': 'CLI', 'ci': False, 'pyversion': '3.8.8', 'samcliVersion': '1.46.0', 'awsProfileProvided': False, 'debugFlagProvided': True, 'region': '', 'commandName': 'sam init', 'duration': 20652, 'exitReason': 'AppTemplateUpdateException', 'exitCode': 1}}]}
2022-04-15 14:45:08,429 | HTTPSConnectionPool(host='aws-serverless-tools-telemetry.us-west-2.amazonaws.com', port=443): Read timed out. (read timeout=0.1)
Error: Unstable state when updating repo. Check that you have permissions to create/delete files in C:\Users\xxx\AppData\Roaming\AWS SAM directory or file an issue at https://github.com/aws/aws-sam-cli/issues`

**Comment by Tristanwhack** [comment]
Issue persists on SAM version 1.48.0

**Comment by dcyprianrw** [comment]
Issue persists on SAM version 1.53.0

**Comment by viniciusknob** [comment]
SAM CLI, version 1.53.0, Windows 10

I had the same issue, and it was solved with these steps:

1. Uninstall AWS SAM;
2. Restart computer;
3. Install AWS SAM again.

**Comment by guoliang** [comment]
still issue in 1.53.0

**Comment by RustyLobster** [comment]
same here. still an issue with `SAM CLI, version 1.53.0`

**Comment by jbrook10** [comment]
Same issue for me.

* Env: Windows 10, `SAM CLI, version 1.53.0`

```
2022-07-21 16:26:49 [INFO]: Command: (not started) [C:\Program Files\Amazon\AWSSAMCLI\bin\sam.cmd init --name Fantasy.Nfl.Lifetime.Lambda --no-interactive --dependency-manager cli-package --runtime dotnet6 --app-template hello-world]
2022-07-21 16:26:54 [INFO]: 
Cloning from https://github.com/aws/aws-sam-cli-app-templates (process may take a moment)

2022-07-21 16:27:08 [INFO]: Error: Unstable state when updating repo. Check that you have permissions to create/delete files in C:\Users\xxx\AppData\Roaming\AWS SAM directory or file an issue at https://github.com/aws/aws-sam-cli/issues
```

**Comment by janivimal** [comment]
Facing same issue, gives permission error,  even though running commands under admin mode:

**_Cloning from https://github.com/aws/aws-sam-cli-app-templates (process may take a moment)
Error: Unstable state when updating repo. Check that you have permissions to create/delete files in C:\Users\xxxxx\AppData\Roaming\AWS SAM directory or file an issue at https://github.com/aws/aws-sam-cli/issues_**

**Comment by allannvictor** [comment]
Same issue here too.

Windows 10, SAM CLI, version 1.53.0

`Cloning from https://github.com/aws/aws-sam-cli-app-templates (process may take a moment)
Error: Unstable state when updating repo. Check that you have permissions to create/delete files in C:\Users\ardt\AppData\Roaming\AWS SAM directory or file an issue at https://github.com/aws/aws-sam-cli/issues`

**Comment by hawflau** [comment]
Hey @kannan-kiwitech @Tristanwhack @guoliang @RustyLobster @jbrook10 @janivimal @allannvictor, I believe the issue you are seeing is the same as #4031. You have to opt-in to enable LongPath support in Windows: https://docs.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=powershell#enable-long-paths-in-windows-10-version-1607-and-later

TLDR: Run the following as administrator in Powershell:
```
New-ItemProperty -Path "HKLM:\SYSTEM\CurrentControlSet\Control\FileSystem" `
-Name "LongPathsEnabled" -Value 1 -PropertyType DWORD -Force
```

**Comment by allannvictor** [comment]
@hawflau It's worked! Thanks!

**Comment by janivimal** [comment]
@hawflau - running this command did the trick ! Thank you for sharing this !

**Comment by SumantSrivastava** [comment]
> Hey @kannan-kiwitech @Tristanwhack @guoliang @RustyLobster @jbrook10 @janivimal @allannvictor, I believe the issue you are seeing is the same as #4031. You have to opt-in to enable LongPath support in Windows: https://docs.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=powershell#enable-long-paths-in-windows-10-version-1607-and-later
> 
> TLDR: Run the following as administrator in Powershell:
> 
> ```
> New-ItemProperty -Path "HKLM:\SYSTEM\CurrentControlSet\Control\FileSystem" `
> -Name "LongPathsEnabled" -Value 1 -PropertyType DWORD -Force
> ```

worked for me too.. thanks!

**Comment by abinshihab** [comment]
@hawflau  Thank you it is working with me

**Comment by keithwenzler** [comment]
Yes, works!
